File: src/page/KeyHandler.java
package page;

import common.CommonStatic.FakeKey;

import java.awt.event.KeyEvent;
import java.util.Map;
import java.util.TreeMap;

import static java.awt.event.KeyEvent.*;

public abstract class KeyHandler extends Page implements FakeKey {

	private static final long serialVersionUID = 1L;

	protected Map<Integer, Integer> press = new TreeMap<>();

	protected int[][] slots = new int[][] { { VK_Q, VK_W, VK_E, VK_R, VK_T }, { VK_A, VK_S, VK_D, VK_F, VK_G },
			{ VK_Z, VK_X, VK_C, VK_V, VK_B } };
	protected int[] act = new int[] { VK_1, VK_2, VK_3 };

	protected int[] lock = new int[] { VK_SHIFT };

	protected KeyHandler(Page p) {
		super(p);
	}

	@Override
	public synchronized boolean pressed(int type, int slot) {
		int key = (type == -2 ? lock : type == -1 ? act : slots[type])[slot];
		return press.containsKey(key) && press.get(key) != 2;
	}

	@Override
	public synchronized void remove(int type, int slot) {
		int key = (type == -2 ? lock : type == -1 ? act : slots[type])[slot];
		press.put(key, 2);
	}

	@Override
	protected synchronized void keyPressed(KeyEvent e) {
		press.put(e.getKeyCode(), 0);
	}

	@Override
	protected synchronized void keyReleased(KeyEvent e) {
		int val = e.getKeyCode();
		if (press.containsKey(val) && press.get(val) != 2)
			press.put(e.getKeyCode(), 1);
	}

	protected synchronized void updateKey() {
		for (int val : press.keySet())
			if (press.get(val) == 1)
				press.put(val, 2);
	}

}

File: src/main/java/page/KeyHandler.kt
package page

import common.CommonStatic
import common.CommonStatic.FakeKey
import java.awt.event.KeyEvent
import java.util.*

abstract class KeyHandler protected constructor(p: Page?) : Page(p), FakeKey {
    protected var press: MutableMap<Int, Int> = TreeMap()
    protected var slots = arrayOf(
        intArrayOf(KeyEvent.VK_Q, KeyEvent.VK_W, KeyEvent.VK_E, KeyEvent.VK_R, KeyEvent.VK_T), intArrayOf(
            KeyEvent.VK_A, KeyEvent.VK_S, KeyEvent.VK_D, KeyEvent.VK_F, KeyEvent.VK_G
        ), intArrayOf(KeyEvent.VK_Z, KeyEvent.VK_X, KeyEvent.VK_C, KeyEvent.VK_V, KeyEvent.VK_B)
    )
    protected var act = intArrayOf(KeyEvent.VK_1, KeyEvent.VK_2, KeyEvent.VK_3, KeyEvent.VK_4)
    protected var change = intArrayOf(KeyEvent.VK_A)
    protected var lock = intArrayOf(KeyEvent.VK_SHIFT)
    @Synchronized
    override fun pressed(type: Int, slot: Int): Boolean {
        return if (CommonStatic.getConfig().twoRow) {
            val key = (if (type == -2) lock else if (type == -1) act else slots[type])[slot]
            press.containsKey(key) && press[key] != 1
        } else {
            val key = (if (type == -3) change else if (type == -2) lock else if (type == -1) act else slots[0])[slot]
            press.containsKey(key) && press[key] != 1
        }
    }

    @Synchronized
    override fun remove(type: Int, slot: Int) {
        if (CommonStatic.getConfig().twoRow) {
            val key = (if (type == -2) lock else if (type == -1) act else slots[type])[slot]
            press[key] = 3
        } else {
            val key = (if (type == -3) change else if (type == -2) lock else if (type == -1) act else slots[0])[slot]
            press[key] = 3
        }
    }

    @Synchronized
    override fun keyPressed(e: KeyEvent) {
        super.keyPressed(e)
        press[e.keyCode] = 2
    }

    @Synchronized
    override fun keyReleased(e: KeyEvent) {
        val `val` = e.keyCode
        if (press.containsKey(`val`) && press[`val`] != 1) press[e.keyCode] = 3
    }

    @Synchronized
    protected fun updateKey() {
        for (`val` in press.keys) {
            if (press[`val`] == 2) press[`val`] = 0 else if (press[`val`] == 3) press[`val`] = 1
        }
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
